lib/puppetserver/ca/action/list.rb

```diff
@@ -71,24 +82,59 @@ def run(input)
           puppet = Config::Puppet.parse(config)
           return 1 if Errors.handle_with_usage(@logger, puppet.errors)
 
-          filter_names = certnames.any? \
-            ? lambda { |x| certnames.include?(x['name']) }
-            : lambda { |x| true }
+          if certnames.any?
+            filter_names = lambda { |x| certnames.include?(x['name']) }
+          else
+            filter_names = lambda { |x| true }
+          end
 
           all_certs = get_all_certs(puppet.settings).select { |cert| filter_names.call(cert) }
           requested, signed, revoked = separate_certs(all_certs)
           missing = certnames - all_certs.map { |cert| cert['name'] }
 
-          (all || certnames.any?) \
-            ? output_certs_by_state(requested, signed, revoked, missing)
-            : output_certs_by_state(requested)
+          if (all || certnames.any?)
+            output_certs_by_state(all, output_format, requested, signed, revoked, missing)
+          else
+            output_certs_by_state(all, output_format, requested)
+          end
+
+          return missing.any? ? 1 : 0
+        end
 
-          return missing.any? \
-            ? 1
-            : 0
+        def output_certs_by_state(all, output_format, requested, signed = [], revoked = [], missing = [])
+          if output_format == 'json'
+            output_certs_json_format(all, requested, signed, revoked, missing)
+          else
+            output_certs_text_format(requested, signed, revoked, missing)
+          end
         end
 
-        def output_certs_by_state(requested, signed = [], revoked = [], missing = [])
+        def output_certs_json_format(all, requested, signed, revoked, missing)
+          grouped_cert = {}
+
+          if all
+            grouped_cert = { "requested" => requested,
+                             "signed" => signed,
+                             "revoked" => revoked }.to_json
+            @logger.inform(grouped_cert)
+          else
+            grouped_cert["requested"] = requested unless requested.empty?
+            grouped_cert["signed"] = signed unless signed.empty?
+            grouped_cert["revoked"] = revoked unless revoked.empty?
+            grouped_cert["missing"] = missing unless missing.empty?
+
+            # If neither the '--all' flag or the '--certname' flag was passed in
+            # then we will display the requested certificates. Otherwise, display
+            # the classes that are in grouped_cert
+            if grouped_cert.empty?
+              @logger.inform({ "requested" => requested }.to_json)
```

Comment:
I'm a little confused about this.. If we've reached this case, `requested` must be empty, right? I'm not understanding why we need this conditional at all.. The caller of this function seems to cover only passing in `requested` certs when `--all` nor `--certname` was passed, so I think you'd still get the desired result without this `if` block.

This also doesn't seem to match what we do for the `text` mode - when everything is empty, we say that, so I would expect the JSON result to just be an empty hash?

Reply:
Ah Justin wanted the case if we passed no args to `list`, we would output whatever that is currently in `requested`.

Reply:
but isn't that taken care of by line 98, by just not passing anything for the other states?
https://github.com/puppetlabs/puppetserver-ca-cli/pull/80/files#diff-805a8aad50ad3c6c816ce6c5607c63d12a5ab8c37a5994877013fd87f88c3a74R98 ?

Reply:
https://github.com/puppetlabs/puppetserver-ca-cli/blob/58c082b7fe8ac77bd8cf267795577e25b17bfc85/lib/puppetserver/ca/cli.rb#L94

Reply:
My thought process was that at any given moment if `grouped_cert` was empty, that meant that we didn't pass in any arguments, otherwise there has to be at least something in the classes, hence why I did that. I see what you meant and I wasn't sure of how I could accommodate to Justin's suggestion so I landed on that :)

Reply:
I think this takes care of the case when someone passed no flag, _and_ there are no requested certs. Makes it so we will output `{ requested: [] }` instead of just `{}`, which I think makes sense to me. Might want to update the comment to describe the specific circumstance when we expect to hit this branch.

Reply:
^ What I tried to explain in a very poorly way.

Reply:
let me see if I can lay this out to un-confuse us both.. 😅 

- for the `--all` case, we pass params for all states and always output all keys, even if some states are empty ✅ 
- for the `--certnames` case, we pass params for all states, but only output keys for non-empty states ✅ 
  - I don't think it's possible for `grouped_cert` to be empty in this case, since anything we don't recognize will land in `missing` ✅ 
- for the no flags case, we only pass a param for `requested`, so the rest will be empty ✅ 
  - the only way for `grouped_cert` to be empty is if there are no `requested` certs - it sounds like Justin was saying we should still return `{ "requested" => [] }` in that case, as opposed to an empty JSON object? 
 
I think I understand now haha, though not totally sure if I agree with that behavior, since the text version just says "No certificates to list".

Reply:
All of your bullet points are correct! At least that's what I understand from Justin's suggestions.

Reply:
I could go either way. I always find myself a little confused about what the behavior is when no flag is passed, and I think this is actually clearer than what the text version does, so I'm kind of in favor of it.

Reply:
ah yeah, that's fair. I'm good with leaving it, but agree that the comment could be clarified to describe exactly how we get here.

Reply:
oh I see you did that already 😅

Reply:
I got y'all lol